crates/hir-ty/src/method_resolution.rs

```diff
@@ -952,7 +954,21 @@ fn iterate_method_candidates_with_autoref(
         )
     };
 
-    iterate_method_candidates_by_receiver(receiver_ty, first_adjustment.clone())?;
+    // If `is_reborrowing` is true, it means that there is a reborrow in progress, so we should
+    // directly go to the autoref step.
+    if !*is_reborrowing {
+        let without_autoref =
+            iterate_method_candidates_by_receiver(receiver_ty, first_adjustment.clone());
+        if without_autoref.is_break()
+            && matches!(receiver_ty.value.kind(Interner), TyKind::Ref(_, _, _))
+        {
+            // if `&&T` and `&T` are both applicable, we should prefer `&*&&T` to `&*&T`, so this
+            // flag exists
```

Comment:
I am somewhat confused by why we need this weird flag at all. This feels very wrong to have.

Reply:
This test is the problem:
```
#[test]
fn method_resolution_by_value_before_autoref() {
    check_types(
        r#"
trait Clone { fn clone(&self) -> Self; }
struct S;
impl Clone for S {}
impl Clone for &S {}
fn test() { (S.clone(), (&S).clone(), (&&S).clone()); }
          //^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ (S, S, &S)
"#,
    );
}
```
We should continue for the reborrow only one time, and this flag serve that propose. Without it, we will get `(S, S, S)`.

Reply:
Ah, now I see what we want to do here. We need to set the `autoref` field of the `ReceiverAdjustments` `first_adjustment` as well as increment the `autoderefs` field once for this `iterate_method_candidates_by_receiver` call when the `receiver_ty` is a reference

Reply:
That then also mimics what rustc does here